### PR TITLE
Include file list when editing split commit description

### PR DESCRIPTION
**Pull request by ilyagr**
Example:

        JJ: Enter commit description for the first part (parent).
        Better split commit message

        JJ: This part contains the following changes:
        JJ:   M src/formatter.rs
        JJ:
        JJ: Lines starting with "JJ: " (like this one) will be removed.

### CODE REVIEW

**Comment by martinvonz** [comment]
It seems that we shouldn't restrict this feature to `split`. It would be useful to always include that information when editing the description, right? I think Git and Mercurial do that. Are you interested in expanding the scope of this PR to do that? It's fine if you're not.

**Comment by ilyagr** [comment]
I thought about it. Expanding this to all messages requires more refactoring. I may or may not be interested in doing that later. So far, `split` is the one case where I got confused without seeing which files are in which part.

**Comment by martinvonz** [comment]
Could you add a test for this? Search for `set_up_fake_editor` in the tests to see how to do it.

**Comment by ilyagr** [comment]
I can look into making a test, but not today. What's the most similar existing test?

**Comment by martinvonz** [comment]
> I can look into making a test, but not today. What's the most similar existing test?

`test_commit_with_editor()` and `test_describe()` both check that the initial contents of the text file is as expected. The former also modifies the description after checking it. I don't know if you'll want to do that.

File: src/commands.rs [hunk 1]
```diff
@@ -2882,6 +2882,31 @@ don't make any changes, then the operation will be aborted.",
...
+    intro: &str,
+    overall_commit_description: &str,
+    diff_iter: TreeDiffIterator,
+) -> Result<String, CommandError> {
+    let result1 = format!(
+        "JJ: {intro}\n{}\nJJ: This part contains the following changes:\n",
+        overall_commit_description,
+    );
+
+    let mut formatter = PlainTextFormatter::new(vec![]);
```
Comment (martinvonz) [inline src/commands.rs:2896]: If you leave the buffer in a local variable and pass in a writer to that, you don't need to reach inside (undo the change the `formatter.rs`). Like this:

```
let mut diff_summary_bytes = vec![];
let mut formatter = PlainTextFormatter::new(&mut buf);
show_diff_summary(&mut formatter, workspace_command, diff_iter)?;
let diff_summary = String::from_utf8(diff_summary_bytes).unwrap();
```

File: src/commands.rs [hunk 2]
```diff
@@ -2882,6 +2882,31 @@ don't make any changes, then the operation will be aborted.",
     Ok(())
 }
 
+fn _description_template_for_cmd_split(
+    workspace_command: &WorkspaceCommandHelper,
+    intro: &str,
+    overall_commit_description: &str,
+    diff_iter: TreeDiffIterator,
+) -> Result<String, CommandError> {
+    let result1 = format!(
```
Comment (martinvonz) [inline src/commands.rs:2891]: These variable names (`result1`, `resultt2`, `jj_result2`) are not very descriptive. We can probably even remove this one by doing it at the end, i.e. with a `format!()` as the last expression in this function.

File: src/commands.rs [hunk 3]
```diff
@@ -2913,11 +2938,20 @@ don't make any changes, then the operation will be aborted.
...
             ui,
             tx.base_repo(),
-            &("JJ: Enter commit description for the first part.\n".to_string()
-                + commit.description()),
+            &_description_template_for_cmd_split(
+                &workspace_command,
+                "Enter commit description for the first part (parent).",
+                commit.description(),
+                base_tree.diff(&middle_tree, &EverythingMatcher),
+            )?,
```
Comment (martinvonz) [inline src/commands.rs:2954]: This expression is long enough that I think it would be clearer if it's extracted to a variable (same below, of course).

File: src/commands.rs (hunk 1 above)
Comment (ilyagr) [reply src/commands.rs:2896]: This works. I didn't expect `PlainTextFormatter<&mut Vec<u8>>` to be a valid type, but it seems to be.

File: src/commands.rs [hunk 4]
```diff
@@ -2971,6 +2971,31 @@ don't make any changes, then the operation will be aborted.",
...
+    show_diff_summary(&mut formatter, workspace_command, diff_iter)?;
+
+    let mut result2 = String::default();
+    std::str::from_utf8(formatter.output())
+        .map(|r| result2.push_str(r))
+        .unwrap_or_else(|err| result2.push_str(&format!("Error displaying diff: {err}")));
+    let jj_result2 = result2
+        .split('\n')
+        .map(|line| "JJ:   ".to_string() + line)
+        .join("\n");
```
[the remark this answers is not in the document]
Comment (ilyagr) [reply src/commands.rs:2906]: That's very nice!

File: src/commands.rs [hunk 5]
```diff
@@ -2882,6 +2882,31 @@ don't make any changes, then the operation will be aborted.",
...
+        overall_commit_description,
+    );
+
+    let mut formatter = PlainTextFormatter::new(vec![]);
+    show_diff_summary(&mut formatter, workspace_command, diff_iter)?;
+
+    let mut result2 = String::default();
+    std::str::from_utf8(formatter.output())
+        .map(|r| result2.push_str(r))
+        .unwrap_or_else(|err| result2.push_str(&format!("Error displaying diff: {err}")));
```
[the remark this answers is not in the document]
Comment (ilyagr) [reply src/commands.rs:2902]: I take it, diffs are always valid UTF8? I didn't assume that because I could imagine some weird cases:

- Weird filenames that aren't valid UTF8. I guess you're saying these are not allowed in jj repos?

- If we ever include an option to show file content, and the files contain crazy content. I guess we don't show a diff in this case, even for long diffs?

[the remark this answers is not in the document]
Comment (martinvonz) [reply src/commands.rs:2902]: > * Weird filenames that aren't valid UTF8. I guess you're saying these are not allowed in jj repos?

Exactly.

> If we ever include an option to show file content, and the files contain crazy content. I guess we don't show a diff in this case, even for long diffs?

Never say never, but I expect the `diff_summary()` function used here to never include content diffs (we have other functions for that).

**Review by martinvonz — APPROVED** [review]
